### ISSUE TITLE
How topic with different types of events should be structured?

### ISSUE THREAD

**Issue by agoretsky** [asyncapi/topic-definition#5]
Hello! I have a kafka topic with several types of related events, for example:
```
must.user_product_collection.1.event.user.watched_movie
must.user_product_collection.1.event.user.watched_tv_episodes
```

then how should I call the topic? now it's `must.user_product_collection.1.event`

And I also would like to say some words about AsyncAPI project itself: Excellent work, thanks!

**Comment by fmvilas** [comment]
Hey @agoretsky!

Thanks for your words. I guess your first two approaches are almost there, but I'd change it to be like this:

```
must.user_product_collection.1.event.user.movie.watched
must.user_product_collection.1.event.user.tv_episodes.watched
```

So the last word is always a verb in past tense because the event already happened, right? Now `user`, `movie` and `tv_episodes` are [resources and subresources](https://github.com/asyncapi/topic-definition#resources-and-sub-resources).

I'm assuming `must` is the name of the company or project.

Does it sound good?

**Comment by agoretsky** [comment]
@fmvilas good correction, thanks! 
But I also would like know is it ok to have a single topic called `must.user_product_collection.1.event`, and messages with `'message_type'` field which can have value `'must.user_product_collection.1.event.user.movie.watched'` or `'must.user_product_collection.1.event.user.tv_episodes.watched'` inside this single topic?

**Comment by fmvilas** [comment]
Ah sorry, didn't get your question then.

**TL;DR:** I'd not use a single topic to subscribe to multiple messages because you're moving the "routing logic" to your app when you already have it on the message broker. If you need to subscribe to multiple topics at once you can use wildcards and make your service subscribe to `must.user_product_collection.1.event.user.*.watched`.

----

Suppose you have the following setup: You have a service in charge of processing all the `watched` events for your TV episodes and movies.

##### Scenario 1:
Suddenly, people get crazy watching TV episodes and you're having problems processing "movie watched" events. They get slow because your service is receiving a high rate of "TV episode watched" events and doesn't have time to process "movie watched" events at a nice pace. You can't split this service into 2 pieces because it's listening for **both events in a single topic**, so your only choice here is to scale massively and pray that people will not get crazy watching movies 😉 

If you'd have subscribed to `must.user_product_collection.1.event.user.*.watched` then you could perfectly split your service into 2 services:

* One subscribed to `must.user_product_collection.1.event.user.movie.watched`.
* Another one subscribed to `must.user_product_collection.1.event.user.tv_episode.watched`.

And scale each of them independently.

##### Scenario 2:
Your message looks like this:

```yaml
message_type: movie or episode
title: My awesome movie and/or tv episode
directors: ["john", "jane"]
duration: 120
```

and so your AsyncAPI message definition looks like this:

```yml
movieOrEpisode:
  type: object
  properties:
    message_type:
      type: string
      enum: ["movie", "episode"]
    title:
      type: string
    ...
```

But then you realize you actually want to include the TV episode's season too. However, the season is not something that makes sense in a movie. So you start doing nasty stuff like leaving it empty or null if it's a movie. Over time both types of messages can evolve in a very different way making it harder for you to preserve a single message for both. And harder to test too!

There's nothing wrong with this approach (except in scenario 1) but I personally don't recommend it. That's exactly the same thing as [function overloading](https://en.wikipedia.org/wiki/Function_overloading). It can be useful sometimes but it's often a sign of bad architecture.

----

Hope it helps!
